### ISSUE TITLE
Model stops learning when using average batch loss instead of sum (Why even use sum?)

### ISSUE THREAD

**Issue by msvensson222** [tensorflow/recommenders#213]
I would like to have the average batch loss, instead of the sum. This because it simplifies a lot of things when performing hyper-parameter tuning in regards to learning-rate, regularization parameters and so on.

To do this, I've simply replaced the default task from the source:
```
    self._loss = loss if loss is not None else tf.keras.losses.CategoricalCrossentropy(
        from_logits=True, reduction=tf.keras.losses.Reduction.SUM)
```

with the loss (`SUM_OVER_BATCH_SIZE `):

```
task = tfrs.tasks.Retrieval(
  loss=tf.keras.losses.CategoricalCrossentropy(
                from_logits=True, reduction=tf.keras.losses.Reduction.SUM_OVER_BATCH_SIZE                           
            ),
  metrics=tfrs.metrics.FactorizedTopK(
      candidates=movies.batch(128).map(movie_model)
  )
)
```

When I do this change in the[ basic retrieval example](https://www.tensorflow.org/recommenders/examples/basic_retrieval), the loss never decreases, thus stops learning and top-k performance obviously goes down the drain. I've played around with learning rates and batch sizes but all to no avail.

Because of this, I have two questions:

1. In almost all of neural-network architectures I've seen, the batch-average loss is the norm - so why not for tensorflow-recommenders models?
2. How can I successfully perform the basic retrieval example with batch-average loss?

**Comment by maciejkula** [comment]
@biteorange could you handle this one?

**Comment by msvensson222** [comment]
> @biteorange could you handle this one?

I would really appreciate some explanation here.

**Comment by biteorange** [comment]
Do you mind share a colab with the above-mentioned modifications? It's quite surprising that the two reductions will have dramatic different results (especially given the Adagrad optimizer).

In the basic retrieval example, I modified the loss and compared the two reduction options, but couldn't reproduce the error.
With the `SUM_OVER_BATCH_SIZE` loss reduction, it actually achieved better train and eval metrics (which is expected).

## With `SUM`

### train log
Epoch 1/3
10/10 [==============================] - 26s 2s/step - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0166 - factorized_top_k/top_10_categorical_accuracy: 0.0358 - factorized_top_k/top_50_categorical_accuracy: 0.1709 - factorized_top_k/top_100_categorical_accuracy: 0.2984 - loss: 66021.3544 - regularization_loss: 0.0000e+00 - total_loss: 66021.3544
Epoch 2/3
10/10 [==============================] - 24s 2s/step - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0238 - factorized_top_k/top_10_categorical_accuracy: 0.0500 - factorized_top_k/top_50_categorical_accuracy: 0.2097 - factorized_top_k/top_100_categorical_accuracy: 0.3442 - loss: 64940.2031 - regularization_loss: 0.0000e+00 - total_loss: 64940.2031
Epoch 3/3
10/10 [==============================] - 24s 2s/step - factorized_top_k/top_1_categorical_accuracy: 0.0016 - factorized_top_k/top_5_categorical_accuracy: 0.0274 - factorized_top_k/top_10_categorical_accuracy: 0.0563 - factorized_top_k/top_50_categorical_accuracy: 0.2250 - factorized_top_k/top_100_categorical_accuracy: 0.3632 - loss: 64442.6257 - regularization_loss: 0.0000e+00 - total_loss: 64442.6257
<tensorflow.python.keras.callbacks.History at 0x7f8623ab07f0>

### test log
5/5 [==============================] - 6s 965ms/step - factorized_top_k/top_1_categorical_accuracy: 1.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0018 - factorized_top_k/top_10_categorical_accuracy: 0.0065 - factorized_top_k/top_50_categorical_accuracy: 0.0740 - factorized_top_k/top_100_categorical_accuracy: 0.1710 - loss: 31917.4342 - regularization_loss: 0.0000e+00 - total_loss: 31917.4342
{'factorized_top_k/top_100_categorical_accuracy': **0.17100000381469727**,
 'factorized_top_k/top_10_categorical_accuracy': 0.006500000134110451,
 'factorized_top_k/top_1_categorical_accuracy': 0.0001500000071246177,
 'factorized_top_k/top_50_categorical_accuracy': 0.07395000010728836,
 'factorized_top_k/top_5_categorical_accuracy': 0.0017500000540167093,
 'loss': 28925.3671875,
 'regularization_loss': 0,
 'total_loss': 28925.3671875}

## With `SUM_OVER_BATCH_SIZE`
### train log
Epoch 1/3
10/10 [==============================] - 26s 2s/step - factorized_top_k/top_1_categorical_accuracy: 0.0027 - factorized_top_k/top_5_categorical_accuracy: 0.0367 - factorized_top_k/top_10_categorical_accuracy: 0.0719 - factorized_top_k/top_50_categorical_accuracy: 0.2553 - factorized_top_k/top_100_categorical_accuracy: 0.3974 - loss: 8.1170 - regularization_loss: 0.0000e+00 - total_loss: 8.1170
Epoch 2/3
10/10 [==============================] - 24s 2s/step - factorized_top_k/top_1_categorical_accuracy: 0.0026 - factorized_top_k/top_5_categorical_accuracy: 0.0365 - factorized_top_k/top_10_categorical_accuracy: 0.0718 - factorized_top_k/top_50_categorical_accuracy: 0.2554 - factorized_top_k/top_100_categorical_accuracy: 0.3974 - loss: 8.1169 - regularization_loss: 0.0000e+00 - total_loss: 8.1169
Epoch 3/3
10/10 [==============================] - 24s 2s/step - factorized_top_k/top_1_categorical_accuracy: 0.0024 - factorized_top_k/top_5_categorical_accuracy: 0.0366 - factorized_top_k/top_10_categorical_accuracy: 0.0720 - factorized_top_k/top_50_categorical_accuracy: 0.2554 - factorized_top_k/top_100_categorical_accuracy: 0.3975 - loss: 8.1168 - regularization_loss: 0.0000e+00 - total_loss: 8.1168
<tensorflow.python.keras.callbacks.History at 0x7f8623d77c18>

### test log
5/5 [==============================] - 5s 941ms/step - factorized_top_k/top_1_categorical_accuracy: 5.0000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0047 - factorized_top_k/top_10_categorical_accuracy: 0.0121 - factorized_top_k/top_50_categorical_accuracy: 0.0917 - factorized_top_k/top_100_categorical_accuracy: 0.1937 - loss: 7.9858 - regularization_loss: 0.0000e+00 - total_loss: 7.9858
{'factorized_top_k/top_100_categorical_accuracy': **0.19374999403953552**,
 'factorized_top_k/top_10_categorical_accuracy': 0.01209999993443489,
 'factorized_top_k/top_1_categorical_accuracy': 0.0005000000237487257,
 'factorized_top_k/top_50_categorical_accuracy': 0.09165000170469284,
 'factorized_top_k/top_5_categorical_accuracy': 0.004699999932199717,
 'loss': 7.887784957885742,
 'regularization_loss': 0,
 'total_loss': 7.887784957885742}

**Comment by msvensson222** [comment]
Thats very weird, since that's exactly what I did and got the results mentioned. The only thing I change in the [retrieval example](https://colab.research.google.com/github/tensorflow/recommenders/blob/main/docs/examples/basic_retrieval.ipynb) is the loss, nothing else. And the results I get are (with `SUM_OVER_BATCH_SIZE`):
## train log
Epoch 1/3
10/10 [==============================] - 24s 2s/step - factorized_top_k/top_1_categorical_accuracy: 1.3750e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0022 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0292 - factorized_top_k/top_100_categorical_accuracy: 0.0587 - loss: 8.9624 - regularization_loss: 0.0000e+00 - total_loss: 8.9624
Epoch 2/3
10/10 [==============================] - 22s 2s/step - factorized_top_k/top_1_categorical_accuracy: 1.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0022 - factorized_top_k/top_10_categorical_accuracy: 0.0053 - factorized_top_k/top_50_categorical_accuracy: 0.0293 - factorized_top_k/top_100_categorical_accuracy: 0.0588 - loss: 8.9624 - regularization_loss: 0.0000e+00 - total_loss: 8.9624
Epoch 3/3
10/10 [==============================] - 22s 2s/step - factorized_top_k/top_1_categorical_accuracy: 1.6250e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0022 - factorized_top_k/top_10_categorical_accuracy: 0.0053 - factorized_top_k/top_50_categorical_accuracy: 0.0293 - factorized_top_k/top_100_categorical_accuracy: 0.0588 - loss: 8.9624 - regularization_loss: 0.0000e+00 - total_loss: 8.9624
<tensorflow.python.keras.callbacks.History at 0x7fdb3e44e198>

## test log
5/5 [==============================] - 5s 918ms/step - factorized_top_k/top_1_categorical_accuracy: 4.5000e-04 - factorized_top_k/top_5_categorical_accuracy: 0.0026 - factorized_top_k/top_10_categorical_accuracy: 0.0054 - factorized_top_k/top_50_categorical_accuracy: 0.0291 - factorized_top_k/top_100_categorical_accuracy: 0.0578 - loss: 8.2762 - regularization_loss: 0.0000e+00 - total_loss: 8.2762
{'factorized_top_k/top_100_categorical_accuracy': 0.05779999867081642,
 'factorized_top_k/top_10_categorical_accuracy': 0.005350000225007534,
 'factorized_top_k/top_1_categorical_accuracy': 0.00044999999227002263,
 'factorized_top_k/top_50_categorical_accuracy': 0.029100000858306885,
 'factorized_top_k/top_5_categorical_accuracy': 0.0026499999221414328,
 'loss': 8.193075180053711,
 'regularization_loss': 0,
 'total_loss': 8.193075180053711}

You can find the notebook [here](https://colab.research.google.com/drive/12cpm5Gx59qb2PUVCmXEoY4ENN6tLXrl7?usp=sharing).

**Comment by biteorange** [comment]
Can you try dramatically increasing the learning rate? (the reasoning is, for the initial steps, the effective gradient is actually divided by batch_size, which is 8192). 

I tried increased the learning rate to 100, and it was working as intended.

**Comment by msvensson222** [comment]
Yes, it worked as you specified by dramatically increasing the learning rate for the example notebook provided. The second part of my initial question was why even use the batch-sum in the first place? As mentioned,  it simplifies a lot of things when performing hyper-parameter tuning in regards to learning-rate, regularization parameters and so on.
